source/agora/test/EnrollmentManager.d

```diff
@@ -75,4 +75,12 @@ unittest
     // check if nodes contains enrollment data previously sent.
     nodes.each!(node =>
         retryFor(node.hasEnrollment(enroll.utxo_key) == true, 5.seconds));
+
+    // tests for revealing a preimage
+    auto preimage_1 = node_1.getNodePreimage(10000);
+    assert(preimage_1 == PreimageInfo.init);
+    preimage_1 = node_1.getNodePreimage(1000);
+    node_1.sendPreimage(preimage_1);
+    auto preimage_2 = node_2.getValidatorPreimage(preimage_1.enroll_key);
+    assert(preimage_1 == preimage_2);
```

Comment:
This is not black box testing. Adding 2 special methods means you aren't really testing the API.
A better test would be to send a pre-image and check if the (X) next block contains it.
Additionally you could filter the `sendPreimage` function of a block to see if it propagates correctly even in the absence of the preimage in one block (but that depends on Henry's PR)/

Reply:
My original approach is not to save pre-images into a block. It's stored only in the validator set table. A pre-image can be restored from another nodes if the database has been crashed. In that case, another API endpoint like getPreimage is needed with the getNodePreimage and getValidatorPreimage removed. Anyway the two functions could be confusing.

Reply:
There is another approach about saving pre-images. It's to store the source value of the random seed encrypted with a node's private key into the Enrollment structure. With that, if the database crashed, the source value of the random seed used for enrollment can be restored from the chain.